File: Sources/ActivityTracker/Tracking/TrackingCoordinator.swift
import Foundation
import AppKit
import Combine

enum TrackingStatus: String {
    case tracking = "Tracking"
    case idle = "Idle"
    case paused = "Paused"
}

/// Central brain: owns the current tracking state, wires together the individual trackers,
/// and is the single place that decides "what app interval is open right now".
final class TrackingCoordinator: ObservableObject {
    static let shared = TrackingCoordinator()

    @Published private(set) var status: TrackingStatus = .tracking
    @Published private(set) var currentAppName: String = "—"

    private let store = ActivityStore.shared
    private let settings = AppSettings.shared

    private var appSwitchObserver: AppSwitchObserver!
    private var idleMonitor: IdleMonitor!
    private var browserTracker: BrowserTracker!
    private var windowTitleTracker: WindowTitleTracker!
    private var activityScoreTracker: ActivityScoreTracker!
    private var screenshotManager: ScreenshotManager!
    private var meetingDetector: MeetingDetector!

    /// Row id of whatever app_intervals row is currently open. Shared with the
    /// activity-score and screenshot trackers so their rows can reference it.
    private(set) var currentIntervalId: Int64?

    private var isManuallyPaused = false

    /// State captured at the moment tracking goes idle, so that on resume we can
    /// offer to reattribute the idle stretch back to whatever was active before —
    /// see `handleIdleChanged` and the idle-resume prompt.
    private var idleStartDate: Date?
    private var idleIntervalId: Int64?
    private var appNameBeforeIdle: String?

    private init() {
        appSwitchObserver = AppSwitchObserver { [weak self] app in
            self?.handleAppActivated(app)
        }
        idleMonitor = IdleMonitor(thresholdProvider: { [weak self] in
            self?.settings.idleThresholdSeconds ?? 180
        }, onIdleChanged: { [weak self] isIdle in
            self?.handleIdleChanged(isIdle)
        }, supplementaryIdleSecondsProvider: { [weak self] in
            self?.activityScoreTracker.secondsSinceLastEvent()
        })
        browserTracker = BrowserTracker { [weak self] title, url, domain in
            self?.store.updateOpenInterval(windowTitle: title, url: url, domain: domain)
        }
        windowTitleTracker = WindowTitleTracker { [weak self] title in
            self?.store.updateOpenInterval(windowTitle: title)
        }
        activityScoreTracker = ActivityScoreTracker { [weak self] minute, count, score in
            self?.store.recordActivityScore(minuteTimestamp: minute, eventCount: count, score: score, appIntervalId: self?.currentIntervalId)
            self?.screenshotManager.noteLatestScore(score)
        }
        screenshotManager = ScreenshotManager(intervalIdProvider: { [weak self] in self?.currentIntervalId })
        meetingDetector = MeetingDetector()
    }

    func start() {
        // Seed with whatever app is frontmost right now.
        if let app = NSWorkspace.shared.frontmostApplication {
            handleAppActivated(app)
        }
        meetingDetector.start() // independent of frontmost-app tracking — can run alongside anything
        appSwitchObserver.start()
        idleMonitor.start()
        activityScoreTracker.start()
        applyScreenshotSetting()
    }

    func applyScreenshotSetting() {
        if settings.screenshotsEnabled {
            screenshotManager.start(intervalMinutes: settings.screenshotIntervalMinutes)
        } else {
            screenshotManager.stop()
        }
    }

    // MARK: - Menu bar controls

    func togglePause() {
        isManuallyPaused.toggle()
        if isManuallyPaused {
            status = .paused
            browserTracker.stop()
            windowTitleTracker.stop()
            // Close whatever was open and log nothing further — pausing is a
            // deliberate choice, not something to measure, so the paused stretch
            // should be a clean gap in the data, not its own tracked category
            // (unlike idle, which the app detects on its own and does log).
            // Closing here matters regardless: leaving the prior interval open
            // would let it get silently credited with the whole paused duration
            // the moment tracking resumes (starting a new interval closes the
            // old one by stamping it with "now").
            store.closeCurrentInterval()
            currentIntervalId = nil
            // Drop any in-flight idle bookkeeping — if you pause mid-idle, that
            // stretch is now fully covered by the gap above, and letting stale
            // idle-start/interval-id state survive into a later, unrelated
            // idle-resume prompt could misattribute time that already includes
            // the paused stretch.
            idleStartDate = nil
            idleIntervalId = nil
            appNameBeforeIdle = nil
        } else {
            status = .tracking
            if let app = NSWorkspace.shared.frontmostApplication {
                handleAppActivated(app)
            }
        }
    }

    // MARK: - Internal state transitions

    private func handleAppActivated(_ app: NSRunningApplication) {
        guard !isManuallyPaused else { return }
        guard status != .idle else { return } // resumed by handleIdleChanged instead

        let name = app.localizedName ?? app.bundleIdentifier ?? "Unknown"
        currentAppName = name
        browserTracker.stop()
        windowTitleTracker.stop()

        if settings.isExcluded(appName: name) {
            store.closeCurrentInterval()
            currentIntervalId = nil
            return
        }

        currentIntervalId = store.startInterval(appName: name, windowTitle: nil, url: nil, domain: nil, isIdle: false)

        if BrowserTracker.isSupportedBrowser(bundleIdentifier: app.bundleIdentifier) {
            browserTracker.start(bundleIdentifier: app.bundleIdentifier!)
        } else {
            windowTitleTracker.start(pid: app.processIdentifier)
        }
    }

    private func handleIdleChanged(_ isIdle: Bool) {
        guard !isManuallyPaused else { return }
        if isIdle {
            status = .idle
            browserTracker.stop()
            windowTitleTracker.stop()
            // Only worth remembering if there was a real, trackable app before this —
            // not the "—" placeholder (no app seen yet) or an excluded app.
            appNameBeforeIdle = (currentAppName != "—" && currentIntervalId != nil) ? currentAppName : nil
            idleStartDate = Date()
            currentIntervalId = store.startInterval(appName: "Idle", windowTitle: nil, url: nil, domain: nil, isIdle: true)
            idleIntervalId = currentIntervalId
        } else {
            let idleDuration = idleStartDate.map { Date().timeIntervalSince($0) } ?? 0
            let intervalToReattribute = idleIntervalId
            let previousApp = appNameBeforeIdle
            idleStartDate = nil
            idleIntervalId = nil
            appNameBeforeIdle = nil

            status = .tracking
            if let app = NSWorkspace.shared.frontmostApplication {
                handleAppActivated(app)
            }

            if settings.idlePromptEnabled,
               idleDuration >= settings.idlePromptThresholdSeconds,
               let intervalToReattribute, let previousApp {
                IdleResumePromptController.shared.show(idleDuration: idleDuration, previousAppName: previousApp) { [weak self] in
                    Log.info("Idle stretch (\(Int(idleDuration))s) reattributed to \(previousApp)")
                    self?.store.reattributeAsTracked(intervalId: intervalToReattribute, appName: previousApp)
                }
            }
        }
    }

    func shutdown() {
        store.closeCurrentInterval()
        meetingDetector.stop()
        store.closeAllOpenMeetingSessions()
    }
}

File: windows/src/ActivityTracker.Core/Tracking/TrackingCoordinator.cs
using System;
using ActivityTracker.Core.Database;
using ActivityTracker.Core.Models;

namespace ActivityTracker.Core.Tracking;

public enum TrackingStatus { Tracking, Idle, Paused }

/// <summary>
/// Central brain: owns the current tracking state, wires together the
/// individual trackers, and is the single place that decides "what app
/// interval is open right now" — the Windows port of the macOS app's
/// TrackingCoordinator. Must be started from the UI thread (the trackers rely
/// on a running Win32 message loop for their hooks).
/// </summary>
public sealed class TrackingCoordinator
{
    private static readonly Lazy<TrackingCoordinator> _shared = new(() => new TrackingCoordinator());
    public static TrackingCoordinator Shared => _shared.Value;

    public event Action<TrackingStatus>? StatusChanged;
    public TrackingStatus Status { get; private set; } = TrackingStatus.Tracking;
    public string CurrentAppName { get; private set; } = "—";

    /// <summary>
    /// Fired when returning from an idle stretch longer than
    /// IdlePromptThresholdSeconds — the UI layer shows a popup asking whether
    /// to keep the idle time as work, and calls the supplied action if so.
    /// </summary>
    public Action<TimeSpan, string, Action>? IdleResumePromptRequested;

    private readonly ActivityStore _store = ActivityStore.Shared;
    private readonly AppSettings _settings = AppSettings.Shared;

    private readonly ForegroundAppTracker _appTracker;
    private readonly IdleMonitor _idleMonitor;
    private readonly ActivityScoreTracker _activityScoreTracker;
    private readonly MeetingDetector _meetingDetector;

    private long? _currentIntervalId;
    private bool _isManuallyPaused;

    private DateTime? _idleStartDate;
    private long? _idleIntervalId;
    private string? _appNameBeforeIdle;

    private TrackingCoordinator()
    {
        _appTracker = new ForegroundAppTracker(HandleAppActivated);
        _idleMonitor = new IdleMonitor(() => _settings.IdleThresholdSeconds, HandleIdleChanged);
        _activityScoreTracker = new ActivityScoreTracker((minute, count, score) =>
            _store.RecordActivityScore(minute, count, score, _currentIntervalId));
        _meetingDetector = new MeetingDetector();
    }

    public void Start()
    {
        _appTracker.Start();
        _idleMonitor.Start();
        _activityScoreTracker.Start();
        _meetingDetector.Start();
    }

    public void TogglePause()
    {
        _isManuallyPaused = !_isManuallyPaused;
        if (_isManuallyPaused)
        {
            SetStatus(TrackingStatus.Paused);
            // Close whatever was open and log nothing further — pausing is a
            // deliberate choice, not something to measure (unlike idle, which
            // the app detects on its own and does log as its own category).
            _store.CloseCurrentInterval();
            _currentIntervalId = null;
            // Drop any in-flight idle bookkeeping — see the equivalent comment
            // in the macOS TrackingCoordinator for why this matters.
            _idleStartDate = null;
            _idleIntervalId = null;
            _appNameBeforeIdle = null;
        }
        else
        {
            SetStatus(TrackingStatus.Tracking);
            _appTracker.ReportCurrentForeground();
        }
    }

    private void HandleAppActivated(string appName, string? windowTitle)
    {
        if (_isManuallyPaused) return;
        if (Status == TrackingStatus.Idle) return; // resumed by HandleIdleChanged instead

        CurrentAppName = appName;

        if (_settings.IsExcluded(appName))
        {
            _store.CloseCurrentInterval();
            _currentIntervalId = null;
            return;
        }

        _currentIntervalId = _store.StartInterval(appName, windowTitle, null, null, false);
    }

    private void HandleIdleChanged(bool isIdle)
    {
        if (_isManuallyPaused) return;

        if (isIdle)
        {
            SetStatus(TrackingStatus.Idle);
            _appNameBeforeIdle = (CurrentAppName != "—" && _currentIntervalId != null) ? CurrentAppName : null;
            _idleStartDate = DateTime.Now;
            _currentIntervalId = _store.StartInterval("Idle", null, null, null, true);
            _idleIntervalId = _currentIntervalId;
        }
        else
        {
            var idleDuration = _idleStartDate.HasValue ? DateTime.Now - _idleStartDate.Value : TimeSpan.Zero;
            var intervalToReattribute = _idleIntervalId;
            var previousApp = _appNameBeforeIdle;
            _idleStartDate = null;
            _idleIntervalId = null;
            _appNameBeforeIdle = null;

            SetStatus(TrackingStatus.Tracking);
            _appTracker.ReportCurrentForeground();

            if (_settings.IdlePromptEnabled
                && idleDuration.TotalSeconds >= _settings.IdlePromptThresholdSeconds
                && intervalToReattribute.HasValue && previousApp != null)
            {
                IdleResumePromptRequested?.Invoke(idleDuration, previousApp, () =>
                {
                    Log.Info($"Idle stretch ({(int)idleDuration.TotalSeconds}s) reattributed to {previousApp}");
                    _store.ReattributeAsTracked(intervalToReattribute.Value, previousApp);
                });
            }
        }
    }

    public void Shutdown()
    {
        _store.CloseCurrentInterval();
        _appTracker.Stop();
        _idleMonitor.Stop();
        _activityScoreTracker.Stop();
        _meetingDetector.Stop();
        _store.CloseAllOpenMeetingSessions();
    }

    private void SetStatus(TrackingStatus status)
    {
        Status = status;
        StatusChanged?.Invoke(status);
    }
}
